### PR TITLE
T26 - LIST<T>

### PR DESCRIPTION
**Pull request by Ajay07Metamation**

### CODE REVIEW

File: Training/Program.cs [hunk 1]
```diff
@@ -1,7 +1,111 @@
...
+      if (index < 0 || index > count) throw new IndexOutOfRangeException ();
+      if (count == arrayList.Length) Array.Resize (ref arrayList, Capacity * 2);
+      for (int i = index; i < count; i++) arrayList[i] = arrayList[i + 1];
+      count--;
+   }
+
+   /// <summary>Method to print the list</summary>
+   public void PrintList () {
+      Console.Write ("List : ");
+      for (int i = 0; i < count; i++) {
```
Comment (priyarathna) [inline Training/Program.cs:105]: Can be re-written as: Console.Write (string.Join (' ', arrayList))

Comment (Ajay07Metamation) [reply Training/Program.cs]: All changes done in the latest commit

File: Training/Program.cs [hunk 2]
```diff
@@ -1,7 +1,111 @@
...
+         if (index < 0 || index > Capacity) throw new IndexOutOfRangeException ();
+         return marrayList[index];
+      }
+      set {
+         if (index < 0 || index > Capacity) throw new IndexOutOfRangeException ();
+         marrayList[index] = value;
+      }
+   }
+
+   /// <summary>Method Add to add an element in the list at the end</summary>
```
Comment (priyarathna) [inline Training/Program.cs:53]: Update the comment lines.

File: Training/Program.cs [hunk 3]
```diff
@@ -1,7 +1,111 @@
...
+list.PrintList (); ;
+list.RemoveAt (2);
+list.PrintList ();
+list.Clear ();
+list.PrintList ();
+
+/// <summary>Creating a class MyList with underlying structure as array</summary>
+class MyList<T> {
+   // Private variables for creating array and int variable for count
+   T[] marrayList;
```
Comment (priyarathna) [inline Training/Program.cs:28]: Should be marrayList => mArrayList. Apply it wherever necessary.

File: Training/Program.cs [hunk 4]
```diff
@@ -1,7 +1,111 @@
...
+      if (a == null) throw new ArgumentNullException ();
+      if (mcount == Capacity) Array.Resize (ref marrayList, Capacity * 2);
+      marrayList[mcount++] = a;
+   }
+
+   /// <summary>Method remove to remove a particular element</summary>
+   /// <param name="a">The element in the list to be removed</param>
+   /// <exception cref="InvalidOperationException"></exception>
+   public bool Remove (T a) {
+      if (Array.IndexOf (marrayList, a) == -1) throw new InvalidOperationException ();
```
Comment (priyarathna) [inline Training/Program.cs:65]: Array.IndexOf is used twice. Store it in a local variable and use that for the next time.

File: Training/Program.cs [hunk 5]
```diff
@@ -1,7 +1,111 @@
...
+      if (index < 0 || index > mcount) throw new IndexOutOfRangeException ();
+      if (mcount == marrayList.Length) Array.Resize (ref marrayList, Capacity * 2);
+      for (int i = index; i < mcount; i++) marrayList[i] = marrayList[i + 1];
+      mcount--;
+   }
+
+   /// <summary>Method to print the list</summary>
+   public void PrintList () {
+      Console.Write ("List : ");
+      for (int i = 0; i < mcount; i++) {
```
Comment (priyarathna) [inline Training/Program.cs:106]: Flower-brace is not needed for single statements.

File: Training/Program.cs [hunk 6]
```diff
@@ -1,7 +1,111 @@
...
+// Create a list (MyList<T>) with underlying structure as array using property,methods and private variables
+
+// TEST CASE
+
+MyList<int> list = new ();
+list.Add (1);
+list.Add (2);
+list.Add (3);
+list.Add (4);
+list.Add (5);
```
Comment (priyarathna) [inline Training/Program.cs:11]: If I insert a test case: list.Remove (3) in line 11, the program gets crashed.

Comment (Ajay07Metamation) [reply Training/Program.cs]: Done all changes in the latest commit

File: Training/Program.cs [hunk 7]
```diff
@@ -1,7 +1,123 @@
...
+internal class Program {
+   private static void Main (string[] args) {
+      MyList<int> list = new ();
+      list.Add (1);
+      list.Add (2);
+      list.Add (3);
+      list.Add (4);
+      list.PrintList ();
+      WriteLine (list.Capacity);
+      list.Remove (3);
```
Comment (priyarathna) [inline Training/Program.cs:16]: Replace this as line.Remove (4). Program gets crashed.

File: Training/Program.cs [hunk 8]
```diff
@@ -1,7 +1,123 @@
...
+      for (int i = elementPositon; i < mCount - 1; i++) mArrayList[i] = mArrayList[i + 1];
+      mCount--;
+      return true;
+   }
+
+   /// <summary>Clears the list</summary>
+   /// <exception cref="InvalidOperationException"></exception>
+   public void Clear () {
+      if (mArrayList.Length <= 0) throw new InvalidOperationException ();
+      Array.Clear (mArrayList, 0, mCount - 1);
```
Comment (priyarathna) [inline Training/Program.cs:89]: 'Clear' should clear the entire list right? But it's working as expected.

File: Training/Program.cs [hunk 9]
```diff
@@ -1,7 +1,123 @@
...
+      list.Add (6);
+      list.Add (7);
+      list.Add (8);
+      list.PrintList ();
+      WriteLine (list.Count);
+      list.Insert (4, 9);
+      list.PrintList ();
+      list.Remove (4);
+      list.PrintList (); ;
+      list.RemoveAt (2);
```
Comment (priyarathna) [inline Training/Program.cs:28]: Replace this line as: list.RemoveAt (7). Program gets crashed.

**Review by priyarathna — CHANGES_REQUESTED** [review]

File: Training/Program.cs [hunk 10]
```diff
@@ -1,7 +1,133 @@
...
+      mArrayList[mCount++] = element;
+   }
+
+   /// <summary>Remove a particular element given</summary>
+   /// <param name="a">The element in the list to be removed</param>
+   /// <exception cref="InvalidOperationException"></exception>
+   public bool Remove (T element) {
+      var elementPositon = Array.IndexOf (mArrayList, element);
+      if (elementPositon == -1) throw new InvalidOperationException ();
+      if (elementPositon != mCount - 1)
```
Comment (priyarathna) [inline Training/Program.cs:85]: This is not needed. Please check.

File: Training/Program.cs [hunk 11]
```diff
@@ -1,7 +1,133 @@
...
+      mArrayList[index] = element;
+      mCount++;
+   }
+
+   /// <summary>Removes the element at a particular index</summary>
+   /// <param name="index">The index at which the element is to be removed</param>
+   /// <exception cref="IndexOutOfRangeException"></exception>
+   public void RemoveAt (int index) {
+      if (index < 0 || index > mCount - 1) throw new IndexOutOfRangeException ();
+      if (index != mCount)
```
Comment (priyarathna) [inline Training/Program.cs:117]: When this condition will be false?

**Review by priyarathna — CHANGES_REQUESTED** [review]

**Review by priyarathna — APPROVED** [review]

**Pull request closed without merging**